// twitter/lib/twitter.js
var sys = require('sys');
var oauth = require('oauth');

var _twitterConsumerKey = "Gd40fLpO1e87KbMUSMW2xA";
var _twitterConsumerSecret = "okhSmRfDS8zugfW6RWvKJpka4c6aa71VQ0db9r7siQ4";
var _callbackUrl = "http://192.168.1.101:3000/success";

function consumer(){
	return new oauth.OAuth("https://api.twitter.com/oauth/request_token",
					"https://api.twitter.com/oauth/access_token",
					_twitterConsumerKey,
					_twitterConsumerSecret,
					"1.0A",
					null,
					"HMAC-SHA1");
}

module.exports = {
	getRequestToken: function (requestTokenCallback) {
		consumer().getOAuthRequestToken( function (error, oauthToken, oauthTokenSecret, results){
			if (error){
				console.log("Error getting OAuth request token : " + sys.inspect(error));
			} else {
				console.log(oauthToken);
				console.log(oauthTokenSecret);
				requestTokenCallback(oauthToken,oauthTokenSecret);
			}
		});
	},

	getAuthAccessToken: function (user, oauthVerifier, authAccessCallback) {
		var accessToken = '';
		var accessTokenSecret = '';
	
		consumer().getOAuthAccessToken(user.oauthRequestToken, user.oauthRequestTokenSecret, oauthVerifier, function(error, oauthAccessToken, oauthAccessTokenSecret, results) {
    		console.log('--getAuthAccessToken--');
    		console.log(user.oauthRequestToken);
    		console.log(user.oauthRequestTokenSecret);
    		console.log(oauthVerifier);
    	
	    	if (error) {
		  		accessToken = "159904250-k8zLnOWAo5kEvs5W1sTSBYnvjb6oLSSNMHbJKc6O";
	      		accessTokenSecret= "b5KQOE5AB7DmXWtHhva3mRHi5v3oD9LQp6h2XWACZ0";
	    	} else {
		  		accessToken = oauthAccessToken;
	      		accessTokenSecret= oauthAccessTokenSecret;
	    	}
	
	    	authAccessCallback(accessToken, accessTokenSecret);
		});
	},

	getHomeTimeline: function (site, callback) {
		consumer().get("http://api.twitter.com/1/statuses/friends_timeline.json?count=30", site.access_token, site.access_token_secret, function (error, data, response) {
        	if (error) {
				console.log('we are in error folks');
        	} else {
				var tweetArray = JSON.parse(data);
				var tweets = [];
			
				for (var i=0; i < tweetArray.length; i++) {
					var tweet = {
						user: tweetArray[i].user.name,
						text: tweetArray[i].text,
						profileImageUrl: tweetArray[i].user.profile_image_url,
			        	screenName: tweetArray[i].user.screen_name
					};
					tweets.push(tweet);
				}

         		callback(tweets);
      		}  
		});
	},
	
	getUserStream: function(access_token, access_token_secret, streamCallback) {
		var request = consumer().get("https://userstream.twitter.com/2/user.json", access_token, access_token_secret);
		request.addListener('response', function(response) {
			response.setEncoding('utf8');
			response.addListener('data',  function(chunk) {
			    try {
				    var tweet = JSON.parse(chunk);

				    if (tweet.hasOwnProperty('text')) {
					    var data = {
					    	screenName: tweet.user.screen_name,
					    	text: tweet.text,
					    	profileImageUrl: tweet.user.profile_image_url,
					    	user: tweet.user.name
					    };
						streamCallback(data);
					}
				} catch (ex) {
					console.log(ex);
				}
			});
			response.addListener('end', function() {
				console.log('----end----');
			});
		});
		request.end();
	}
};
